wandb/sdk/launch/runner/aws.py

```diff
@@ -0,0 +1,475 @@
+        raise LaunchError("Failed to login to ECR")
+    return login_process.stdout.decode("utf-8")
+
+
+def merge_aws_tag_with_algorithm_specification(
+    algorithm_specification: Optional[Dict[str, Any]], aws_tag: Optional[str]
+) -> Dict[str, Any]:
+    """
+    AWS Sagemaker algorithms require a training image and an input mode.
+    If the user does not specify the specification themselves, define the spec
+    minimally using these two fields. Otherwise, if they specify the AlgorithmSpecification
+    set the training image if it is not set.
+    """
+    if algorithm_specification is None:
+        return {
+            "TrainingImage": aws_tag,
+            "TrainingInputMode": "File",
+        }
+    elif algorithm_specification.get("TrainingImage") is None:
+        algorithm_specification["TrainingImage"] = aws_tag
+    if algorithm_specification["TrainingImage"] is None:
+        raise LaunchError("Failed determine tag for training image")
+    return algorithm_specification
+
+
+def build_sagemaker_args(
+    launch_project: LaunchProject, account_id: str, aws_tag: Optional[str] = None,
+) -> Dict[str, Any]:
+    sagemaker_args = {}
+    resource_args = launch_project.resource_args
+    sagemaker_args["TrainingJobName"] = (
+        resource_args.get("TrainingJobName") or launch_project.run_id
+    )
+
+    sagemaker_args[
+        "AlgorithmSpecification"
+    ] = merge_aws_tag_with_algorithm_specification(
+        resource_args.get("AlgorithmSpecification"), aws_tag,
+    )
+    # required
+    sagemaker_args["OutputDataConfig"] = resource_args.get(
+        "OutputDataConfig", resource_args.get("output_data_config")
+    )
+    if sagemaker_args["OutputDataConfig"] is None:
+        raise LaunchError(
+            "Sagemaker launcher requires an OutputDataConfig resource argument"
+        )
+
+    sagemaker_args["ResourceConfig"] = resource_args.get(
+        "ResourceConfig", resource_args.get("resource_config")
+    )
+    if sagemaker_args["ResourceConfig"] is None:
+        raise LaunchError(
+            "Sagemaker launcher requires a ResourceConfig resource argument"
+        )
+
+    sagemaker_args["StoppingCondition"] = resource_args.get("StoppingCondition")
+    if sagemaker_args["StoppingCondition"] is None:
+        raise LaunchError(
+            "Sagemaker launcher requires a StoppingCondition resource argument"
+        )
+
+    sagemaker_args["RoleArn"] = get_role_arn(resource_args, account_id)
+
+    # optional
+    sagemaker_args["InputDataConfig"] = resource_args.get(
+        "InputDataConfig", resource_args.get("input_data_config")
+    )
+
+    sagemaker_args["HyperParameters"] = resource_args.get(
+        "HyperParameters", resource_args.get("hyper_parameters")
+    )
+    sagemaker_args["VpcConfig"] = resource_args.get(
+        "VpcConfig", resource_args.get("vpc_config")
+    )
+    sagemaker_args["Tags"] = resource_args.get("Tags", resource_args.get("tags"))
```

Comment:
lol this whole section is brutal i wonder if we should even be enabling passing these in through cli args (ie force all of this through a json config instead)

Reply:
the `resource_args` can come through the launch spec and might differ from run to run. But I have an idea to solve this. I'll just check for the required ones, and convert all other snake_case examples to camelcase.

The question is do we want a `sagemaker` key in resource args to define these? I think so.

Reply:
We can subkey all resource args for sagemaker to a subkey of resource_args called `sagemaker` in that case we may as well drop the CLI arg for resource_args. I can drop that in this PR. I agree that this still isn't optimal